### PR TITLE
Fix Braket-PennyLane plugin in preparation for PennyLane v0.15.0 release

### PR DESCRIPTION
**Pull request by trbromley**
*Issue #, if available:*

Fixes https://github.com/aws/amazon-braket-pennylane-plugin-python/issues/61.

*Context:*

A new version of PennyLane is planned for release next week. This version includes some [changes](https://pennylane.readthedocs.io/en/latest/development/release_notes.html) that impact the Braket-PennyLane device, including:

- The `analytic` keyword argument for devices has been removed. Now, `shots=None` indicates analytic mode. ([PR](https://github.com/PennyLaneAI/pennylane/pull/1079))
- The new Tape-based core is now the default and the old core has been remove ([PR](https://github.com/PennyLaneAI/pennylane/pull/1100)). Consequently, `qml.enable_tape()` is no longer available and should be removed.

The [plugin test matrix](https://github.com/PennyLaneAI/plugin-test-matrix#testing-matrix) shows that this plugin is currently failing when used with the development branch of PennyLane (right-most column). Our objective is to have this column pass. Once the new version of PennyLane has been released, it would then be good to release a new version of this plugin.

*Description of changes:*

This PR fixes the currently broken tests. Most changes involve updating the shots/analytic behaviour and removing `enable_tape()`.

One point for discussion:

For the `BraketAwsQubitDevice`, currently `shots=None` is the default, which is internally converted to a default number of shots set by `AwsDevice.DEFAULT_SHOTS_SIMULATOR` or `AwsDevice.DEFAULT_SHOTS_QPU`. Instead, `shots=0` is used to signify analytic mode (not allowed for QPU devices).

Unfortunately, this is incompatible with the new style in core PennyLane, which uses `shots=None` as the method of selecting analytic mode. The fix in this PR is to keep `shots=None` as the default input (i.e., selecting analytic mode), but _not_ to internally swap that out for a default number set by `AwsDevice.DEFAULT_SHOTS_SIMULATOR` or `AwsDevice.DEFAULT_SHOTS_QPU`. As a consequence, if users try to access a QPU device without specifying the number of shots, they will be met with an error (`"QPU devices require the number of shots to be specified"`).

This is a breaking change for those using hardware devices without explicitly setting `shots`. However, I'd argue that it is better to require users explicitly specify their number of shots (forcing them to think about it), rather than using defaults behind the scenes. On the other hand, we could probably avoid a breaking change if we really want to - e.g., by keeping `shots=None` but swapping out for a default value as before, at a cost of not following core PL.

### CODE REVIEW

File: src/braket/pennylane_plugin/braket_device.py
Comment (trbromley) [inline src/braket/pennylane_plugin/braket_device.py:273]: I wasn't sure if we still needed this check?

**Comment by trbromley** [comment]
Note: there is a failing [integration test](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/257eb99d62122e3a83c380ab240d6e166b2b9257/test/integ_tests/test_batch.py#L131) :thinking: 
`ValueError: ('custom_gradient function expected to return', 1, 'gradients but returned', 12, 'instead.')`

File: test/unit_tests/test_braket_device.py
Comment (albi3ro) [inline test/unit_tests/test_braket_device.py:469]: Should rename the test and update the docstring

```suggestion
def test_local_None_shots():
```

File: test/unit_tests/test_braket_device.py
Comment (albi3ro) [inline test/unit_tests/test_braket_device.py:470]:
```suggestion
    """Tests that the simulator devices are analytic if ``shots`` is specified to be `None`."""
```

**Review by albi3ro — CHANGES_REQUESTED** [review]
On [`braket_deivce.py` Line 75](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/257eb99d62122e3a83c380ab240d6e166b2b9257/src/braket/pennylane_plugin/braket_device.py#L75) we need to bump `pennylane_requires` because of the breaking change with shots behaviour.  

Should we also set `None` as the default shots number for [`BraketQubitDevice`](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/257eb99d62122e3a83c380ab240d6e166b2b9257/src/braket/pennylane_plugin/braket_device.py#L84)?

I left a couple of comments, but nothing major.

File: test/integ_tests/test_batch.py
Comment (trbromley) [reply test/integ_tests/test_batch.py:23]: Ah this is because `shots` is a fixture that is used in the `device` fixture - so this just seemed an easy way to get it to work.

**Review by trbromley — COMMENTED** [review]
Thanks @albi3ro for the review!

Good spot with the `pennylane_requires`! Have updated.

> Should we also set None as the default shots number for BraketQubitDevice?

I don't think it matters since this device is not user facing, it is the parent of the local and remote devices.

**Comment by licedric** [comment]
A tricky issue is that the Braket SDK uses shots=0 to indicate analytic mode, which differs from the upcoming PL practice of using shots=None. Probably worth thinking about which one makes more sense here (or perhaps both?)

**Comment by trbromley** [comment]
> A tricky issue is that the Braket SDK uses shots=0 to indicate analytic mode, which differs from the upcoming PL practice of using shots=None. Probably worth thinking about which one makes more sense here (or perhaps both?)

Thanks @licedric! Good point - this is respected at least internally since we still use `shots=0 if self.analytic else self.shots` when calling `_device.run` and `_device.run_batch()`.

However, I'd also be up for allowing users to input `shots=0` when loading the device and still get analytic mode (hence supporting both styles).

File: src/braket/pennylane_plugin/braket_device.py
Comment (trbromley) [inline src/braket/pennylane_plugin/braket_device.py:85]:
```suggestion
        super().__init__(wires, shots=shots or None)
```
@licedric, this should be sufficient (although maybe hacky) to allow users to set `shots=0` and still get analytic mode.

**Review by albi3ro — DISMISSED** [review]
Thank you for addressing my concerns. I'm updating my review to approval :) 👍

File: src/braket/pennylane_plugin/braket_device.py
Comment (kshitijc) [reply src/braket/pennylane_plugin/braket_device.py:85]: I would be interesting in learning on what the experience is like for other Pennylane integrations. Do we bias towards porting the experience for the dependencies we integrate with, or does Pennylane prefer that the users program with Pennylane specific parameters? Ultimately, pennylane users would want to follow a specific convention, that is either they only refer to the parameter convention for the specific integration (which they might anyways have to do for parameters specific to the integrated device) or, they follow the pennylane style parameters.

**Review by kshitijc — COMMENTED** [review]
* *Please note*: Amazon Braket repositories require that breaking changes follow the guidelines outlined in [CONTRIBUTING.md](https://github.com/aws/amazon-braket-pennylane-plugin-python/blob/main/CONTRIBUTING.md).

* Codecov appears to be failing due to insufficient coverage. Please ensure that the changes have complete test coverage.

File: src/braket/pennylane_plugin/braket_device.py
Comment (licedric) [reply src/braket/pennylane_plugin/braket_device.py:85]: @kshitijc, do you see a problem with allowing both the PennyLane and the Braket convention?

File: test/unit_tests/test_braket_device.py
Comment (licedric) [inline test/unit_tests/test_braket_device.py:476]: Can you add a case where you can create a QPU device? That seems to be at least one reason why coverage is failing.

File: src/braket/pennylane_plugin/braket_device.py
Comment (kshitijc) [reply src/braket/pennylane_plugin/braket_device.py:85]: Trying to support both usually tends to cause issues. However, I think supporting both shots=0/None could probably be less jarring for customers using both PL and Braket. Unless, Pennylane plans on implementing some behavior for shots=0 in the future, I'd be okay with supporting both for now.

File: test/unit_tests/test_braket_device.py
Comment (speller26) [reply test/unit_tests/test_braket_device.py:476]: The coverage is actually failing due to two different lines. Fixed in #63

Comment (licedric) [reply test/unit_tests/test_braket_device.py:476]: Actually not sure this is the issue anymore, since Cody just added https://github.com/aws/amazon-braket-pennylane-plugin-python/pull/63 to try to fix coverage.

**Comment by speller26** [comment]
Coverage has been fixed with #63; everything should be good to go once you merge main into your branch and resolve the conflicts!

File: src/braket/pennylane_plugin/braket_device.py
Comment (speller26) [inline src/braket/pennylane_plugin/braket_device.py:227]: We should still use the correct device-specific default shots; a user should be able to use QPUs without explicity setting `shots`. Is `None` the default for all other PennyLane devices?

File: src/braket/pennylane_plugin/braket_device.py
Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:273]: As with the comment on line 227, device-specific defaults should be used

File: doc/xanadu_theme/static/xanadu.css_t
Comment (speller26) [inline doc/xanadu_theme/static/xanadu.css_t:4]: The current font doesn't look like Product Sans at all; where is it being used?

File: setup.py
Comment (speller26) [inline setup.py:31]: Is this still necessary?

File: doc/xanadu_theme/static/xanadu.css_t
Comment (josh146) [reply doc/xanadu_theme/static/xanadu.css_t:4]: Hey @speller26, this was a bug that we have fixed and updated across all the CSS. This change simply brings the Braket docs inline with the rest of the PL documentation

File: src/braket/pennylane_plugin/braket_device.py
Comment (licedric) [reply src/braket/pennylane_plugin/braket_device.py:227]: +1 on what Cody said, since Braket also has a default shots of 1000 for QPUs. It does make it a bit difficult to set a default for QPUs if None is already taken though.

Comment (licedric) [reply src/braket/pennylane_plugin/braket_device.py:227]: Actually I retract what I said - what is the behavior on other quantum hardware? We may want to match that, because we wouldn't want behavior to differ when running a circuit on e.g. IBM and IonQ.

Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:227]: Different devices have different defaults; [pennylane-forest](https://github.com/PennyLaneAI/pennylane-forest/blob/master/pennylane_forest/qpu.py#L96) has 1000, while [pennylane-ionq](https://github.com/PennyLaneAI/PennyLane-IonQ/blob/master/pennylane_ionq/device.py#L83) and [pennylane-qiskit ](https://github.com/PennyLaneAI/pennylane-qiskit/blob/master/pennylane_qiskit/qiskit_device.py#L115) have defaults of 1024. Of course, we can't use the default devices' `None`, since QPUs will be unusable with the default.

Comment (licedric) [reply src/braket/pennylane_plugin/braket_device.py:227]: One way to add a default would be to remove shots as a named parameter, and then try to grab shots from the kwargs. If it doesn't exist, then set it to default (None for simulators / 1000 for QPUs). That does make the code less self-documenting...

File: setup.py
Comment (josh146) [reply setup.py:31]: Nope! This can be changed to
```suggestion
        "pennylane>=0.15",
```

File: src/braket/pennylane_plugin/braket_device.py
Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:85]: Things are complicated by the SDK [already having](https://github.com/aws/amazon-braket-sdk-python/blob/main/src/braket/aws/aws_device.py#L83) a default value of `None`, which is then used to select the device-specific default value.

Comment (josh146) [reply src/braket/pennylane_plugin/braket_device.py:85]: I suppose, stepping back a bit, this is more an issue with having a single PennyLane device for using the remote Braket devices (`BraketAwsQubitDevice`)? By having separate devices per remote device, sensible defaults could be chosen per remote device.

File: doc/devices/braket_local.rst
Comment (trbromley) [inline doc/devices/braket_local.rst:45]:
```suggestion
You can set ``shots`` to ``None`` (default) to get exact results instead of results calculated from samples.
```

File: src/braket/pennylane_plugin/braket_device.py
Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:227]: A nice way to do this would be to create an enum
```python
class Shots(Enum):
    DEFAULT = auto()
```
And set the default value of `shots` to `Shots.DEFAULT`. Then, the original default selection logic returns, with the tweak that explicitly setting `shots=None` translates to `shots=0` in Braket. In short:

| PennyLane | Braket |
| ----------- | ----------- |
| `shots=Shots.DEFAULT` | `shots=None` |
| `shots=None` | `shots=0` |

Comment (licedric) [reply src/braket/pennylane_plugin/braket_device.py:228]: +1 on Cody's suggestion - it looks like the sanest way to keep the existing defaults in place.

Comment (trbromley) [reply src/braket/pennylane_plugin/braket_device.py:228]: I've just pushed some changes that does the following:

Local device:
- `shots=None` is default
- `shots=None` or `shots=0` is analytic mode

Remote device:
- `shots=None` is default (same as before)
- `device_type=AwsDeviceType.SIMULATOR` and `shots=None` or `shots=0` is analytic mode (so same as local dev)
- `device_type=AwsDeviceType.QPU` and `shots=None` adopts non-analytic mode with `shots=DEFAULT_SHOTS_QPU`
- `device_type=AwsDeviceType.QPU` and `shots=0` raises an error (same as before)

I think this can make sense if we make it clear in the documentation.

Can also consider using an `Enum`, I'd just gone down the above route before I saw the message :+1:

**Comment by trbromley** [comment]
@speller26 @licedric - I made some updates to the shots behaviour - check out [this](https://github.com/aws/amazon-braket-pennylane-plugin-python/pull/62#discussion_r626950644) comment.

File: src/braket/pennylane_plugin/braket_device.py
Comment (licedric) [reply src/braket/pennylane_plugin/braket_device.py:228]: Thanks, Tom. Correct me if I'm wrong, but that seems to be the existing behavior before this PR, right?

I think the `Enum` makes more sense here, because of this case:
> `device_type=AwsDeviceType.QPU` and `shots=None` adopts non-analytic mode with `shots=DEFAULT_SHOTS_QPU`

That is code behavior that doesn't match the rest of PennyLane, and I'd be worried this could cause problems down the line.

Comment (speller26) [reply src/braket/pennylane_plugin/braket_device.py:228]: I agree; for this plugin, `None` should be reserved for analytic results, just like the rest of PennyLane.

Comment (trbromley) [reply src/braket/pennylane_plugin/braket_device.py]: I've now updated to use the `Shots` enum. The behaviour is now:

Local device:
- `shots=None` is default
- `shots=None` or `shots=0` is analytic mode

Remote device:
- `shots=Shots.DEFAULT` is default. This value defers the choice of default to remote device.
- Supports `shots=None` or `shots=0` but raises an error if this is chosen when `AwsDeviceType.QPU`
- Now behaves as the rest of PL in that `shots=None` definitely means analytic mode

If this works for you, it'd be great to have a final round of review.

**Review by speller26 — APPROVED** [review]
chef kiss

**Review by licedric — APPROVED** [review]
Thanks for your patience with the back and forth!

**Pull request merged**
